kernel/src/predicates/mod.rs

```diff
@@ -0,0 +1,652 @@
+            GreaterThan => Self::partial_cmp_scalars(Ordering::Greater, left, right, inverted),
+            GreaterThanOrEqual => Self::partial_cmp_scalars(Ordering::Less, left, right, !inverted),
+            _ => {
+                debug!("Unsupported binary operator: {left:?} {op:?} {right:?}");
+                None
+            }
+        }
+    }
+
+    /// Finishes evaluating a (possibly inverted) variadic operation. See
+    /// [`PredicateEvaluator::finish_eval_variadic`].
+    ///
+    /// The inputs were already inverted by the caller, if needed.
+    ///
+    /// With AND (OR), any FALSE (TRUE) input dominates, forcing a FALSE (TRUE) output.  If there
+    /// was no dominating input, then any NULL input forces NULL output.  Otherwise, return the
+    /// non-dominant value. Inverting the operation also inverts the dominant value.
+    pub(crate) fn finish_eval_variadic(
+        op: VariadicOperator,
+        exprs: impl IntoIterator<Item = Option<bool>>,
+        inverted: bool,
+    ) -> Option<bool> {
+        let dominator = match op {
+            VariadicOperator::And => inverted,
+            VariadicOperator::Or => !inverted,
+        };
+        let result = exprs.into_iter().try_fold(false, |found_null, val| {
+            match val {
+                Some(val) if val == dominator => None, // (1) short circuit, dominant found
+                Some(_) => Some(found_null),
+                None => Some(true), // (2) null found (but keep looking for a dominant value)
+            }
+        });
+
+        match result {
+            None => Some(dominator), // (1) short circuit, dominant found
+            Some(false) => Some(!dominator),
+            Some(true) => None, // (2) null found, dominant not found
+        }
+    }
+}
+
+/// Resolves columns as scalars, as a building block for [`DefaultPredicateEvaluator`].
+pub(crate) trait ResolveColumnAsScalar {
+    fn resolve_column(&self, col: &ColumnName) -> Option<Scalar>;
+}
+
+// Some tests do not actually require column resolution
+#[cfg(test)]
+pub(crate) struct UnimplementedColumnResolver;
+#[cfg(test)]
+impl ResolveColumnAsScalar for UnimplementedColumnResolver {
+    fn resolve_column(&self, _col: &ColumnName) -> Option<Scalar> {
+        unimplemented!()
+    }
+}
+
+// In testing, it is convenient to just build a hashmap of scalar values.
+#[cfg(test)]
+impl ResolveColumnAsScalar for std::collections::HashMap<ColumnName, Scalar> {
+    fn resolve_column(&self, col: &ColumnName) -> Option<Scalar> {
+        self.get(col).cloned()
+    }
+}
+
+/// A predicate evaluator that directly evaluates the predicate to produce an `Option<bool>`
+/// result. Column resolution is handled by an embedded [`ResolveColumnAsScalar`] instance.
+pub(crate) struct DefaultPredicateEvaluator {
+    resolver: Box<dyn ResolveColumnAsScalar>,
+}
+
+impl std::ops::Deref for DefaultPredicateEvaluator {
```

Comment:
I don't love this. We don't really _expect_ an evaluator to be able to deref to a totally different type. I get that this let's us do stuff like:
```rust
let col = self.resolve_column(col)?;
```

below, but it's pretty tricky. I'd generally just prefer to do
```rust
let col = self.resolver.resolve_column(col)?;
```

It's more clear and not really that much more typing.  If we keep it, let's at least add a comment explaining why we implement deref to resolver.

Reply:
Is there any chance of this de-ref failing? If so this is likely not something we should do, I +1 to nick, I think the explicit here isn't so bad that I'd want deref coercion.

Reply:
Yeah, this was overly-clever code. Reverting.